translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,435 @@
+	// separate ResourceSpans instance, containing only that span. All other OC Spans
+	// that have nil Resource field must be placed in one other ResourceSpans instance,
+	// which will gets its Resource field from "td.Resource".
+	//
+	// We will end up with with one or more ResourceSpans like this:
+	//
+	// ResourceSpans           ResourceSpans  ResourceSpans
+	// +-----+-----+---+-----+ +------------+ +------------+
+	// |Span1|Span2|...|SpanM| |Span        | |Span        | ...
+	// +-----+-----+---+-----+ +------------+ +------------+
+
+	// Count the number of spans that have nil Resource and need to be combined
+	// in one slice.
+	combinedSpanCount := 0
+	for _, ocSpan := range td.Spans {
+		if ocSpan == nil {
+			// Skip nil spans.
+			continue
+		}
+		if ocSpan.Resource == nil {
+			combinedSpanCount++
+		}
+	}
+
+	// Allocate a slice for spans that need to be combined into one ResourceSpans.
+	combinedSpans := data.NewSpanSlice(combinedSpanCount)
+	resourceSpans.SetSpans(combinedSpans)
+
+	// Now do the span translation and place them in appropriate ResourceSpans
+	// instances.
+
+	// Index to next available slot in "combinedSpans" slice.
+	i := 0
+	for _, ocSpan := range td.Spans {
+		if ocSpan == nil {
+			// Skip nil spans.
+			continue
+		}
+
+		if ocSpan.Resource == nil {
+			// Add the span to the "combinedSpans". combinedSpans length is equal
+			// to combinedSpanCount. The loop above that calculates combinedSpanCount
+			// has exact same conditions as we have here in this loop.
+			destSpan := combinedSpans[i]
+			i++
+
+			// Convert the span.
+			ocSpanToInternal(destSpan, ocSpan)
+		} else {
+			// This span has a different Resource and must be placed in a different
+			// ResourceSpans instance. Create a separate ResourceSpans item just for this span.
+			separateRS := ocSpanToResourceSpans(ocSpan, td.Node, ocSpan.Resource)
+
+			// Add the newly created ResourceSpans to the final list that we will return.
+			resourceSpanList = append(resourceSpanList, separateRS)
+		}
+	}
+
+	return data.NewTraceData(resourceSpanList)
+}
+
+func ocSpanToResourceSpans(ocSpan *octrace.Span, node *occommon.Node, resource *ocresource.Resource) *data.ResourceSpans {
+	destSpan := data.NewSpan()
+	ocSpanToInternal(destSpan, ocSpan)
+
+	// Create a separate ResourceSpans item just for this span.
+	return data.NewResourceSpans(
+		ocNodeResourceToInternal(node, resource),
+		[]*data.Span{destSpan},
+	)
+}
+
+func ocSpanToInternal(dest *data.Span, src *octrace.Span) {
+	events, droppedEventCount := ocEventsToInternal(src.TimeEvents)
+	links, droppedLinkCount := ocLinksToInternal(src.Links)
+
+	attrs := ocAttrsToInternal(src.Attributes)
+	// Note that ocSpanKindToInternal must be called after ocAttrsToInternal
+	// since it may modify the attributes (remove the attribute which represents the
+	// span kind).
+	kind := ocSpanKindToInternal(src.Kind, src.Attributes)
```

Comment:
Not sure what behavior you want to implement here, if you want this behavior why do you remove the attribute which represents the span kind in `ocSpanKindToInternal`. I think you want to opposite order and not propagate the attribute which represents the span kind in the internal attributes.

Reply:
That's a bug, fixing.